core/src/main/java/io/grpc/internal/SingleProducerSerializingExecutor.java

```diff
@@ -0,0 +1,224 @@
+     *   published even without synchronization, if a reference to the containing object is
+     *   observed. See JLS§17.5.
+     * </ol>
+     */
+    LinkedRunnable next;
+
+    public LinkedRunnable(Runnable runnable) {
+      this.runnable = runnable;
+    }
+
+    public void run() {
+      runnable.run();
+    }
+  }
+
+  /**
+   * Task that actually runs the Runnables. The head of the queue must be set before execution. This
+   * is the only reference to the head of the queue, such that it is the only code that can iterate
+   * over the full queue. It executes entries in the queue until it is empty, sets tail=null, and
+   * then returns. This allows the current worker thread to return to the original pool.
+   */
+  private class TaskRunner implements Runnable {
+    private LinkedRunnable head;
+
+    void schedule(LinkedRunnable newHead) {
+      if (head != null) {
+        throw new AssertionError();
+      }
+      head = newHead;
+      try {
+        executor.execute(this);
+      } catch (Throwable t) {
+        // It is possible that 'newHead' is still queued and tail==newHead. But that will trick
+        // execute() into thinking TaskRunner is running (we assume it isn't) which would prevent
+        // all future Runnables from being run. So we clear all state and propagate the exception so
+        // that if our caller deems it recoverable we won't be permanently stuck.
+        tail = null;
+        head = null; // Prevent retaining garbage
+        Throwables.propagateIfPossible(t);
+        throw new AssertionError(t);
+      }
+    }
+
+    @Override
+    public void run() {
+      LinkedRunnable head = this.head;
+      this.head = null; // Prevent retaining garbage after the method returns
+
+      LinkedRunnable cachedTail = tail;
+      while (true) {
+        try {
+          head.run();
+        } catch (Throwable t) {
+          if (head == tail && tailCompareAndSet(head, null)) {
+            // Queue exhausted. No need to reschedule. Just let exception propagate.
+          } else {
+            head = Preconditions.checkNotNull(head.next, "head.next");
+            try {
+              schedule(head);
+            } catch (Throwable t2) {
+              log.log(Level.SEVERE, "Exception while re-scheduling due to exception", t2);
+            }
+          }
+          Throwables.propagateIfPossible(t);
+          throw new AssertionError(t);
+        }
+        // Do not advance in the queue past tail, even if head.next != null, otherwise we could win
+        // a spectatular race by executing head.next before execute() has CAS tail. Then we would be
+        // forced to spin-loop waiting for execute() to catch up; there would be no more work queued
+        // and we couldn't set tail=null and return, because then execute() would reschedule without
+        // knowing the runnable had already been executed.
+        if (head == cachedTail && head == (cachedTail = tail)) {
```

Comment:
Yes, but not in the normal way. It is more impactful here because of all the writes happening on the other thread (and thus necessary cache coherency). I did benchmark with and without the optimization to see if it was detectable, and I think it was like a ~10% improvement (I consider the benchmarks trash, but the fact it detected the change was informative).